### ISSUE TITLE
I cannot activate the "Bambu Cloud Configuration"

### ISSUE THREAD

**Issue by RobertoCravallo** [greghesp/ha-bambulab#673]
### Describe the bug

I get a message that it is not possible, followed by an Email from BAMBU with a verification code. This I can not enter into the integration set-up. My printer is set correctly.

### To Reproduce

Try to configure "Bambu Cloud Configuration" on (my) system.

### Expected Behaviour

As in all videos I watched, being connected to the BAMBU cloud.

### What device are you using?

P1S

### Diagnostic Output

```shell
Can't add my log.  :-(
```

**Comment by RobertoCravallo** [comment]
[home-assistant_bambu_lab_2024-11-04T09-47-30.972Z.log](https://github.com/user-attachments/files/17616349/home-assistant_bambu_lab_2024-11-04T09-47-30.972Z.log)

**Comment by KingJim1984** [comment]
Hello, I still have the same problem. I like to add my A1 Mini Printer to Home Assitant. But when I enter my email adress and Password, I get a activation code to my email adress, and I have no option to enter this code on home assistent. My account ist based in Europe

![grafik](https://github.com/user-attachments/assets/9dfe5558-b459-41cc-8343-ebcd233e4d8d)

**Comment by kurim** [comment]
It seems Bambulab now sends verification codes to make the security stronger, during authentication process.

Same issue: https://github.com/xperiments-in/xtouch/issues/164

**Comment by Schnuecks** [comment]
Same here cant login anymore, i disabled 2fa but still no login possible.

**Comment by Datastore24Kirill** [comment]
![image](https://github.com/user-attachments/assets/72770510-ea30-4067-a4b7-fd9034759d14)
I also have this problem. I can't connect.

**Comment by kanflo** [comment]
BambuLab has updated their authentication method so we need to figure out how to log on. This coincides with the latest version (3 days ago) of BambuHandy that talks about "new verification steps to better protect your account". While this is A Good Thing(tm) it does cause some hassle. The `v1/user-service/user/login` returns and empty response for me (and no auth token email):

```
{'accessToken': '', 'refreshToken': '', 'expiresIn': 0, 'refreshExpiresIn': 0, 'tfaKey': '', 'accessMethod': '', 'loginType': 'verifyCode'}
```

**Comment by Datastore24Kirill** [comment]
> BambuLab has updated their authentication method so we need to figure out how to log on. This coincides with the latest version (3 days ago) of BambuHandy that talks about "new verification steps to better protect your account". While this is A Good Thing(tm) it does cause some hassle. The `v1/user-service/user/login` returns and empty response for me (and no auth token email):
> 
> 
> 
> ```
> 
> {'accessToken': '', 'refreshToken': '', 'expiresIn': 0, 'refreshExpiresIn': 0, 'tfaKey': '', 'accessMethod': '', 'loginType': 'verifyCode'}
> 
> ```

We are waiting for the decision. If necessary, I am ready to provide a Chinese account, login by phone number and password. Two-factor authentication is disabled.

**Comment by kanflo** [comment]
I noticed @kurim nailed it. Using the access code once, I can now log on using email/password. It might be a bit of a hassle to get the access code into this integration from HASS and we will see how often it is needed.

**Comment by t0nyz0** [comment]
I agree @kanflo I was able to get it working using his method with this test harness I built. Thanks @kurim 

```
import requests
import json

# Prompt the user for their Bambu Lab username and password
bambuUsername = input("Enter your Bambu Lab username: ")
bambuPassword = input("Enter your Bambu Lab password: ")

# Define the User-Agent and other common headers
userAgent = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/103.0.0.0 Safari/537.36"
headers = {
    "Content-Type": "application/json",
    "User-Agent": userAgent,
    "Accept": "application/json, text/plain, */*",
    "Accept-Language": "en-US,en;q=0.9",
    "Origin": "https://bambulab.com",
    "Referer": "https://bambulab.com/"
}

# Perform the login request with custom headers
auth_payload = {
    "account": bambuUsername,
    "password": bambuPassword,
    "apiError": ""
}
auth_response = requests.post("https://bambulab.com/api/sign-in/form", headers=headers, json=auth_payload)

# Print the auth response for debugging
print("1 - Auth login response:", auth_response.text)

# Check if authentication was successful
auth_json = auth_response.json()
if not auth_json.get("success"):
    print("Authentication failed, attempting code verification")

    # Send the verification code request
    send_code_payload = {
        "email": bambuUsername,
        "type": "codeLogin"
    }
    send_code_response = requests.post("https://api.bambulab.com/v1/user-service/user/sendemail/code", headers=headers, json=send_code_payload)
    if send_code_response.status_code == 200:
        print("Verification code sent successfully. Check your email.")
    else:
        print("Failed to send verification code.")
        print("Response:", send_code_response.text)
        exit(1)

    if auth_json.get("loginType") == "verifyCode":
        verifyCode = input("Enter your access code: ")
        verify_payload = {
            "account": bambuUsername,
            "code": verifyCode
        }
        api_response = requests.post("https://api.bambulab.com/v1/user-service/user/login", headers=headers, json=verify_payload)
        print("2 - API response:", api_response.text)

apiToken = api_response.json()
if apiToken:
    token = apiToken.get("accessToken")

if not token:
    print("Failed to extract token")
    exit(1)

# Perform the API request to fetch the image with custom headers
headers["Authorization"] = f"Bearer {token}"
api_response = requests.get("https://api.bambulab.com/v1/user-service/my/tasks", headers=headers)

# Print the API response for debugging
print("API to fetch info:", api_response.text)

# Check if the API request was successful
if api_response.status_code != 200:
    print("API request failed")
    exit(1)

# Extract and display the relevant information
api_json = api_response.json()
if api_json:
    hits = api_json.get("hits", [{}])[0]
    imageUrl = hits.get("cover")
    modelTitle = hits.get("title")
    modelWeight = hits.get("weight")
    modelCostTime = hits.get("costTime")
    totalPrints = api_json.get("total")
    deviceName = hits.get("deviceName")
    deviceModel = hits.get("deviceModel")
    bedType = hits.get("bedType")

    # Output the results
    print("Image URL:", imageUrl)
    print("Model Title:", modelTitle)
    print("Model Weight:", modelWeight)
    print("Model Cost Time:", modelCostTime)
    print("Total Prints:", totalPrints)
    print("Device Name:", deviceName)
    print("Device Model:", deviceModel)
```

**Comment by AdrianGarside** [comment]
Thanks. With the above it looks like I should be able to add MFA support to the authentication flow pretty easily.

**Comment by AdrianGarside** [comment]
For folk hitting this - do you have MFA actively enabled on your Bambu account or is this just a new one-time MFA being rolled out globally?

**Comment by t0nyz0** [comment]
I dont have MFA turned on @AdrianGarside 
I believe this "extra security verification code" is rolling out to all users eventually.

MFA seems a little more complex, the initial 'sign-in' post returns:
`'loginType' ='tfa'` instead of `'loginType' ='verifyCode'`

When I post the 6 digit code from my MFA device to:
`https://api.bambulab.com/v1/user-service/user/login`
it says its incorrect. However, as we know this works with the verification code.

**Comment by justenuf2b** [comment]
> Hitting the issue, no 2fa setup on the account

Same. 2-Step Verification is turned off.

**Comment by Datastore24Kirill** [comment]
Same. 2-Step Verification is turned off.

**Comment by AdrianGarside** [comment]
This isn't working for me. I'm getting cloudflare blocked at the first step:
auth_response = requests.post("https://bambulab.com/api/sign-in/form", headers=headers, json=auth_payload)

**Comment by AdrianGarside** [comment]
Ok, that seems to be related to my on the fly conversion to httpx. Running your code exactly as is, got me the expected verifyCode response.

Nope, the same httpx converted code works fine outside of the integration. It just fails when run inside the integration.

**Comment by Datastore24Kirill** [comment]
> Ok, that seems to be related to my on the fly conversion to httpx. Running your code exactly as is, got me the expected verifyCode response.
> 
> 
> 
> Nope, the same httpx converted code works fine outside of the integration. It just fails when run inside the integration.
> 
> 

What to do when two-factor authentication is turned off and authorization takes place by phone number and SMS.

**Comment by t0nyz0** [comment]
@AdrianGarside I got MFA (TFA) working as well, here is the updated code.

```python
import requests
import json

# Prompt the user for their Bambu Lab username and password
bambuUsername = input("Enter your Bambu Lab username: ")
bambuPassword = input("Enter your Bambu Lab password: ")

# Define the User-Agent and other common headers
user_agent = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/103.0.0.0 Safari/537.36"
headers = {
    "Content-Type": "application/json",
    "User-Agent": user_agent,
    "Accept": "application/json, text/plain, */*",
    "Accept-Language": "en-US,en;q=0.9",
    "Origin": "https://bambulab.com",
    "Referer": "https://bambulab.com/"
}

# Perform the login request with custom headers
auth_payload = {
    "account": bambuUsername,
    "password": bambuPassword,
    "apiError": ""
}
auth_response = requests.post("https://bambulab.com/api/sign-in/form", headers=headers, json=auth_payload)

# Print the auth response for debugging
print("Step 1. Auth login response:", auth_response.text)

# Check if authentication was successful
auth_json = auth_response.json()
if not auth_json.get("success"):
    print("Authentication failed, attempting code verification")

    # Send the verification code request
    send_code_payload = {
        "email": bambuUsername,
        "type": "codeLogin"
    }

    if auth_json.get("loginType") == "verifyCode":
        send_code_response = requests.post("https://api.bambulab.com/v1/user-service/user/sendemail/code", headers=headers, json=send_code_payload)
        if send_code_response.status_code == 200:
            print("Verification code sent successfully. Check your email.")
        else:
            print("Failed to send verification code.")
            print("Response:", send_code_response.text)
            exit(1)

        verify_code = input("Enter your access code: ")
        verify_payload = {
            "account": bambuUsername,
            "code": verify_code
        }
        api_response = requests.post("https://api.bambulab.com/v1/user-service/user/login", headers=headers, json=verify_payload)
        print("Step 2 - API verify code response:", api_response.text)

        api_token = api_response.json()
        if api_token:
            token = api_token.get("accessToken")

        if not token:
            print("Failed to extract token")
            exit(1)

    elif auth_json.get("loginType") == "tfa":
        tfa_auth = auth_json.get("tfaKey")
        tfa_code = input("Enter your MFA access code: ")
        verify_payload = {
            "tfaKey": tfa_auth,
            "tfaCode": tfa_code
        }
        print("payload: ", verify_payload)
        api2_response = requests.post("https://bambulab.com/api-sign-in/tfa", headers=headers, json=verify_payload)
        print("Step 2 - API MFA response:", api2_response.text)

        cookies = api2_response.cookies.get_dict()
        token_from_tfa = cookies.get("token")
        print("tokenFromTFA:", token_from_tfa)

        token = token_from_tfa

        if not token:
            print("Failed to extract token")
            exit(1)

else:
    # If authentication was successful in the first attempt, get the token from the JSON response
    token = auth_json.get("accessToken")

if not token:
    print("Unable to authenticate or verify. Exiting...")
    exit(1)

# Perform the API request to fetch the image with custom headers
headers["Authorization"] = f"Bearer {token}"
api_response = requests.get("https://api.bambulab.com/v1/user-service/my/tasks", headers=headers)

# Print the API response for debugging
print("API to fetch info:", api_response.text)

# Check if the API request was successful
if api_response.status_code != 200:
    print("API request failed")
    exit(1)

# Extract and display the relevant information
api_json = api_response.json()
if api_json:
    hits = api_json.get("hits", [{}])[0]
    image_url = hits.get("cover")
    model_title = hits.get("title")
    model_weight = hits.get("weight")
    model_cost_time = hits.get("costTime")
    total_prints = api_json.get("total")
    device_name = hits.get("deviceName")
    device_model = hits.get("deviceModel")
    bed_type = hits.get("bedType")

    # Output the results
    print("Image URL:", image_url)
    print("Model Title:", model_title)
    print("Model Weight:", model_weight)
    print("Model Cost Time:", model_cost_time)
    print("Total Prints:", total_prints)
    print("Device Name:", device_name)
    print("Device Model:", device_model)
    print("Bed Type:", bed_type)

```

**Comment by AdrianGarside** [comment]
Awesome! I still need to work out why this code is refusing to work when run inside the integration. There should be no difference but cloudflare is somehow not liking it when it’s run from within home assistant.

**Comment by AdrianGarside** [comment]
I can reproduce the issue running python inside the home assistant docker manually - it gets blocked by cloudfare while the same commands run on the host do not. Host has python 3.11.2 while my HA docker instance has 3.12.4. So far that's the only difference I've been able to see.

**Comment by AdrianGarside** [comment]
httpx works fine on the host but is bust inside the integration. requests works for both so I guess I'll switch back to requests for at least the authentication flow - but I was seeing issues with the slicer settings accessing using a pre-existing auth token and it's using httpx now too.

**Comment by AdrianGarside** [comment]
https://github.com/greghesp/ha-bambulab/releases/tag/v2.0.28-code1

**Comment by RobertoCravallo** [comment]
> I noticed @kurim nailed it. Using the access code once, I can now log on using email/password.

I have no idea what you mean and how you did it. I believe there will be a lot of need to implemnt this in the integration, because like me, most of us are just users and heavily rely on the skills of others...

**Comment by RobertoCravallo** [comment]
> do you have MFA actively enabled on your Bambu account

No, no MFA.

**Comment by RobertoCravallo** [comment]
[t0nyz0](https://github.com/t0nyz0) Thank you for your work on this!!!!

**Comment by Schnuecks** [comment]
seems to work, the verification code email came twice, if this matters for something. great work guys.

**Comment by RobertoCravallo** [comment]
> Alpha release hot off the presses:

Yupp, working now, THANK YOU!!! My camera still only shows "idle".

**Comment by AdrianGarside** [comment]
Yep I was seeing two emails pretty much every time in my testing. I’m definitely only making a request for one - maybe the request isn’t even necessary?

**Comment by Datastore24Kirill** [comment]
For authorization by phone number of China it did not work
![image](https://github.com/user-attachments/assets/f075e6b5-e2d7-43b0-961f-925eec55d5ba)

**Comment by AdrianGarside** [comment]
Can you please enable debug logs in your configuration.yaml and provide debug logs for that?
```logger:
  default: info
  logs:
    custom_components.bambu_lab: debug```

**Comment by AdrianGarside** [comment]
> Yep I was seeing two emails pretty much every time in my testing. I’m definitely only making a request for one - maybe the request isn’t even necessary?

Yep, not necessary. The code is sent automatically.

**Comment by Datastore24Kirill** [comment]
`2024-11-05 17:33:35.924 ERROR (MainThread) [custom_components.bambu_lab] Failed to connect with error code ('list index out of range',)`

This is the only mention of the component BambuLab.

![image](https://github.com/user-attachments/assets/e6f8eeb1-71b9-4b25-85d5-ae7f38e5b9e4)

**Comment by AdrianGarside** [comment]
This should have the full 2FA support enabled but I haven't tested it yet:
https://github.com/greghesp/ha-bambulab/releases/tag/v2.0.28-code2

**Comment by AdrianGarside** [comment]
> `2024-11-05 17:33:35.924 ERROR (MainThread) [custom_components.bambu_lab] Failed to connect with error code ('list index out of range',)`
> 
> This is the only mention of the component BambuLab.

You don't have debug logging enabled. Did you restart you home assistant after enabling that in the configuration.yaml file?

**Comment by daften** [comment]
This might be a stupid question, but I updated to the code3 version but it doesn't work. I guess I need to re-authenticate, but can't find how to do this. Can somebody give instructions? Or should I uninstall and re-install the hacs extension.

**Comment by AdrianGarside** [comment]
How doesn't it work? Any debug logs?

code3 has the support to reauthenticate in the config flow - but it'll skip that if the existing authentication code it has still works:
![image](https://github.com/user-attachments/assets/b8c47ca0-4e0a-4c58-b101-6a0a811e6d46)

**Comment by daften** [comment]
I went through configuration again, and it works indeed, so just didn't think about that flow, my bad. I said it was potentially a stupid question :)

Thanks so much for the quick reply and quick fixes!

**Comment by alptunga** [comment]
Does v2.0.28-code3 also work for those doing the initial setup or only those previously using the integration? I am unsuccessful with initial setup.

**Comment by AdrianGarside** [comment]
Primarily initial setup. Folks already using the integration already have a saved auth Token that doesn't expire for a year.

**Comment by AdrianGarside** [comment]
Closing this issue now. Please log a new issue for any remaining issues (such as potentially not being able to login in China).
